core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1158,74 +1158,89 @@ public int getNumberOfPinnedFiles() {
   public void delete(AlluxioURI path, DeleteOptions options) throws IOException,
       FileDoesNotExistException, DirectoryNotEmptyException, InvalidPathException,
       AccessControlException {
+    List<Long> blocksToDelete;
     Metrics.DELETE_PATHS_OPS.inc();
     try (JournalContext journalContext = createJournalContext();
          LockedInodePath inodePath = mInodeTree.lockFullInodePath(path, InodeTree.LockMode.WRITE)) {
       mPermissionChecker.checkParentPermission(Mode.Bits.WRITE, inodePath);
       mMountTable.checkUnderWritableMountPoint(path);
-      deleteAndJournal(inodePath, options, journalContext);
+      blocksToDelete = deleteAndJournal(inodePath, options, journalContext);
     }
+    mBlockMaster.removeBlocks(blocksToDelete, true /* delete */);
   }
 
   /**
    * Deletes a given path.
    * <p>
    * Writes to the journal.
    *
+   * This method does not delete blocks. Instead, it returns the blocks so that they can be deleted
+   * after the path deletion journal entry has been written. We cannot delete blocks earlier because
+   * the path deletion may fail, leaving us with files containing deleted blocks.
+   *
    * @param inodePath the path to delete
    * @param deleteOptions the method options
    * @param journalContext the journal context
+   * @return the list of deleted block ids
```

Comment:
NIT: this and others, sound like the blocks were already deleted, not that they should be deleted later.